crates/maple_cli/src/tools/ripgrep/jsont.rs

```diff
@@ -0,0 +1,100 @@
+use serde::Deserialize;
+
+use super::stats::Stats;
+
+#[derive(Deserialize, Debug)]
+#[serde(tag = "type", content = "data")]
+#[serde(rename_all = "snake_case")]
+pub enum Message {
+    Begin(Begin),
+    End(End),
+    Match(Match),
+    Context(Context),
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Begin {
+    pub path: Data,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct End {
+    pub path: Data,
+    pub binary_offset: Option<u64>,
+    pub stats: Stats,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Match {
+    pub path: Data,
+    pub lines: Data,
+    pub line_number: Option<u64>,
+    pub absolute_offset: u64,
+    pub submatches: Vec<SubMatch>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Context {
+    pub path: Data,
+    pub lines: Data,
+    pub line_number: Option<u64>,
+    pub absolute_offset: u64,
+    pub submatches: Vec<SubMatch>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct SubMatch {
+    #[serde(rename = "match")]
+    pub m: Data,
+    pub start: usize,
+    pub end: usize,
+}
+
+/// Copied from https://github.com/BurntSushi/ripgrep/blob/ce4b587055/crates/printer/src/jsont.rs#L82
+///
+/// Data represents things that look like strings, but may actually not be
+/// valid UTF-8. To handle this, `Data` is serialized as an object with one
+/// of two keys: `text` (for valid UTF-8) or `bytes` (for invalid UTF-8).
+///
+/// The happy path is valid UTF-8, which streams right through as-is, since
+/// it is natively supported by JSON. When invalid UTF-8 is found, then it is
+/// represented as arbitrary bytes and base64 encoded.
+#[derive(Clone, Debug, Hash, PartialEq, Eq, Deserialize)]
+#[serde(untagged)]
+pub enum Data {
+    Text {
+        text: String,
+    },
+    Bytes {
+        #[serde(deserialize_with = "from_base64")]
+        bytes: Vec<u8>,
+    },
+}
+
+impl Data {
+    pub fn text(&self) -> String {
+        match self {
+            Self::Text { text } => text.to_string(),
+            Self::Bytes { bytes } => String::from_utf8_lossy(bytes).to_string(),
+        }
+    }
```

Comment:
`String::from_utf8_lossy(bytes).to_string()` allocates one more string than you need. Also, `Cow` here should be more effective, especially if most of `Data` would be text:

```rust
pub fn text(&self) -> Cow<str> {
    match self {
        Self::Text { text } => text.as_str().into(),
        Self::Bytes { bytes } => String::from_utf8_lossy(bytes).into(),
    }
}
```

Then, if you'd need a `String` from `Cow`, it has a `Cow::into_owned` function.